File: lib/muxpad/palette.rb
# frozen_string_literal: true

require "io/console"

module Muxpad
  # An interactive item shown in the palette. +state+ is the human label and
  # +state_kind+ drives its colour (:running, :idle, :finished, :available,
  # :disabled, :unavailable).
  Item = Data.define(:token, :section, :name, :description, :command, :directory, :state, :state_kind)

  # A terminal-native command palette: sectioned, fuzzy-searchable, scrollable.
  # The view layer only renders and reads keys; all launch behaviour lives in
  # Application. #select returns [action, token] (action is "enter", "tab" or
  # "ctrl-r") or nil when cancelled. #choose returns a chosen token or nil.
  class Palette
    # Non-list lines: title, blank, search, blank, blank, command, directory, hint.
    CHROME = 8
    MIN_LIST = 3
    RIGHT_PAD = 2 # blank columns kept clear on the right edge
    NAME_MIN = 10
    NAME_MAX = 24
    RUNNING_SECTION = "RUNNING" # split into the left sidebar rather than the launch list
    SIDEBAR_WIDTH = 20

    RESET = "\e[0m"
    BOLD = "\e[1m"
    DIM = "\e[2m"
    REVERSE = "\e[7m"
    HEADER = "\e[1;35m" # bold magenta
    HIDE_CURSOR = "\e[?25l"
    SHOW_CURSOR = "\e[?25h"

    STATE_COLOR = {
      running: "\e[32m", finished: "\e[33m", available: "\e[32m",
      disabled: DIM, unavailable: "\e[31m", idle: DIM
    }.freeze

    def initialize(input: $stdin, output: $stdout, prompt: "Muxpad")
      @input = input
      @output = output
      @prompt = prompt
    end

    def select(items, section_order:)
      raise Error, "the Muxpad palette requires an interactive terminal" unless @input.tty?
      @running = items.select { |item| item.section == RUNNING_SECTION }
      @launch_items = items.reject { |item| item.section == RUNNING_SECTION }
      @section_order = section_order
      @name_width = name_width(@launch_items)
      @focus = :launch
      @run_cursor = 0
      @query = +""
      refilter
      interact { |key| handle_select_key(key) }
    end

    def choose(options, title:)
      raise Error, "the Muxpad palette requires an interactive terminal" unless @input.tty?
      @choices = options
      @cursor = 0
      render_choice(title)
      @input.raw do
        loop do
          case read_key
          when :up then move_choice(-1)
          when :down then move_choice(1)
          when :enter then return @choices[@cursor]&.first
          when :escape, :cancel then return nil
          else next
          end
          render_choice(title)
        end
      end
    ensure
      teardown
    end

    private

    def interact
      render
      @input.raw do
        loop do
          result = yield read_key
          return result.first if result
          render
        end
      end
    ensure
      teardown
    end

    # Returns nil to keep looping, or a one-element array wrapping the result to
    # return from #select (so that a nil result can itself be returned).
    def handle_select_key(key)
      case key
      when :up then move(-1)
      when :down then move(1)
      when :left then set_focus(:running)
      when :right then set_focus(:launch)
      when :backspace then edit { @query.chop! }
      when :clear then edit { @query.clear }
      when Array then edit { @query << key.last } # [:char, c]
      when :enter then return [selection("enter")] if current
      when :tab then return [selection("tab")] if @focus == :launch && current
      when :restart then return [selection("ctrl-r")] if @focus == :launch && current
      when :escape, :cancel then return [nil]
      end
      nil
    end

    def selection(action)
      [action, current.token]
    end

    # The sidebar only exists while something is running; otherwise the launch
    # list owns the full width and focus stays put.
    def set_focus(pane)
      return if pane == :running && @running.empty?
      @focus = pane
    end

    # Typing always searches the launch list, so editing the query pulls focus
    # back to it from the sidebar.
    def edit
      @focus = :launch
      yield
      refilter
    end

    # Rebuild the grouped, ranked view for the current query and the flat list of
    # rendered rows (headers, blanks, items) the cursor and scroller walk over.
    def refilter
      ranked = @launch_items.filter_map do |item|
        score = score(item)
        [item, score] if score
      end
      grouped = ranked.group_by { |item, _| item.section }
      @rows = []
      @section_order.each do |section|
        entries = grouped[section]
        next unless entries&.any?
        entries = entries.sort_by.with_index { |(_, score), i| [-score, i] }
        @rows << { kind: :blank } unless @rows.empty?
        @rows << { kind: :header, label: section }
        entries.each { |item, _| @rows << { kind: :item, item: } }
      end
      @selectable = @rows.each_index.select { |i| @rows[i][:kind] == :item }
      @cursor = 0
      @offset = 0
    end

    def current
      return @running[@run_cursor] if @focus == :running
      row = @selectable[@cursor]
      row && @rows[row][:item]
    end

    def move(delta)
      if @focus == :running
        return if @running.empty?
        @run_cursor = (@run_cursor + delta).clamp(0, @running.length - 1)
        return
      end
      return if @selectable.empty?
      @cursor = (@cursor + delta).clamp(0, @selectable.length - 1)
      line = @selectable[@cursor]
      height = list_height
      @offset = line if line < @offset
      @offset = line - height + 1 if line > @offset + height - 1
    end

    def move_choice(delta)
      @cursor = (@cursor + delta).clamp(0, @choices.length - 1)
    end

    # --- scoring -------------------------------------------------------------

    # Subsequence match across name/command/description, weighted in that order
    # so a name hit always outranks a command hit, which outranks a description
    # hit. Returns nil when the query matches no field.
    def score(item)
      return 0 if @query.empty?
      query = @query.downcase
      [[item.name, 3], [item.command, 2], [item.description, 1]].filter_map do |text, weight|
        inner = subsequence_score(text.to_s.downcase, query)
        weight * 100_000 + inner if inner
      end.max
    end

    def subsequence_score(text, query)
      first = nil
      from = 0
      query.each_char do |char|
        index = text.index(char, from)
        return nil unless index
        first ||= index
        from = index + 1
      end
      span = from - 1 - first
      [10_000 - first * 10 - span, 0].max
    end

    # --- rendering -----------------------------------------------------------

    def render
      width = columns - RIGHT_PAD
      lines = ["#{BOLD}  #{@prompt}#{RESET}", "", "  #{BOLD}❯#{RESET} #{@query}", ""]
      lines.concat(body_lines(width))
      lines << ""
      lines.concat(detail_lines(width))
      lines << "  #{DIM}#{hint}#{RESET}"
      paint(lines)
      cursor_to(3, 5 + display_width(@query))
    end

    def hint
      switch = sidebar? ? "←/→ switch · " : ""
      "#{switch}enter launch · tab actions · ctrl-r restart · esc close"
    end

    def sidebar? = @running.any?

    # When something is running, the list region is two columns: the running
    # sidebar on the left and the launch list on the right, joined by a divider
    # that runs the height of the taller column.
    def body_lines(width)
      launch = launch_lines(width)
      return launch unless sidebar?

      launch = launch_lines(width - SIDEBAR_WIDTH - 1)
      side = sidebar_lines(SIDEBAR_WIDTH)
      height = [[launch.length, side.length].max, list_height].min
      (0...height).map do |i|
        "#{side[i] || (" " * SIDEBAR_WIDTH)}#{DIM}│#{RESET}#{launch[i]}"
      end
    end

    # Only the rows that actually exist within the viewport — no blank filler, so
    # the detail strip rises to meet a short list instead of floating below a gap.
    def launch_lines(width)
      last = [@offset + list_height, @rows.length].min
      (@offset...last).map { |i| render_row(@rows[i], width) }
    end

    def render_row(row, width)
      case row[:kind]
      when :blank then ""
      when :header then "  #{HEADER}#{row[:label]}#{RESET}"
      when :item then render_item(row[:item], width, @focus == :launch && current.equal?(row[:item]))
      end
    end

    def render_item(item, width, selected)
      avail = width - 2
      state = truncate(item.state.to_s, [avail / 3, 30].min)
      name = truncate(item.name.to_s, @name_width).ljust(@name_width)
      desc_width = [avail - @name_width - 2 - state.length - 1, 0].max
      desc = truncate(item.description.to_s, desc_width).ljust(desc_width)
      if selected
        # Keep the text within the content width, but extend the highlight bar
        # through the reserved right padding so it reaches the popup edge.
        line = "  #{name}  #{desc} #{state}"
        "#{REVERSE}#{line[0, width].ljust(width + RIGHT_PAD)}#{RESET}"
      else
        color = STATE_COLOR.fetch(item.state_kind, DIM)
        "  #{name}  #{DIM}#{desc}#{RESET} #{color}#{state}#{RESET}"
      end
    end

    # The left sidebar: a RUNNING header followed by each live instance, each as
    # a coloured dot and name, padded to the fixed sidebar width.
    def sidebar_lines(width)
      cells = [fill(" #{HEADER}RUNNING#{RESET}", 8, width)]
      @running.each_with_index do |item, i|
        name = truncate(item.name.to_s, width - 3)
        if @focus == :running && i == @run_cursor
          cells << "#{REVERSE}#{" ● #{name}"[0, width].ljust(width)}#{RESET}"
        else
          cells << fill(" #{STATE_COLOR[:running]}●#{RESET} #{name}", 3 + name.length, width)
        end
      end
      cells
    end

    # Pad a string carrying ANSI codes to a visible +width+ using its known
    # printable length, so colour codes don't throw the column alignment off.
    def fill(text, visible, width)
      visible >= width ? text : text + (" " * (width - visible))
    end

    # The two-line preview: the exact command, and the directory it runs in.
    def detail_lines(width)
      item = current
      return ["", ""] unless item
      command = "  #{DIM}$ #{truncate(item.command.to_s, width - 4)}#{RESET}"
      directory = if item.directory.to_s.empty?
        ""
      else
        "  #{DIM}in #{truncate(abbreviate(item.directory), width - 5)}#{RESET}"
      end
      [command, directory]
    end

    def abbreviate(path)
      home = Dir.home
      path.start_with?("#{home}/") ? path.sub(home, "~") : path
    end

    def render_choice(title)
      width = columns - RIGHT_PAD
      lines = ["#{BOLD}  #{title}#{RESET}", ""]
      @choices.each_with_index do |(_, label), i|
        if i == @cursor
          lines << "#{REVERSE}#{"  #{label}".ljust(width)[0, width]}#{RESET}"
        else
          lines << "  #{label}"
        end
      end
      lines << ""
      lines << "  #{DIM}enter select · esc cancel#{RESET}"
      paint(lines)
      cursor_to(lines.length, 1)
    end

    def paint(lines)
      body = lines.map { |line| "#{line}\e[K" }.join("\r\n")
      @output.print("#{HIDE_CURSOR}\e[H#{body}\e[J")
    end

    def cursor_to(row, col)
      @output.print("\e[#{row};#{col}H#{SHOW_CURSOR}")
      @output.flush
    end

    def teardown
      @output.print("#{SHOW_CURSOR}\e[2J\e[H")
      @output.flush
    end

    # --- key input -----------------------------------------------------------

    def read_key
      char = @input.getch
      case char
      when "\r", "\n" then :enter
      when "\t" then :tab
      when "", "\b" then :backspace
      when "" then :cancel
      when "" then :restart
      when "" then :down
      when "" then :up
      when "" then :clear
      when "\e" then escape_key
      else printable(char)
      end
    end

    def escape_key
      return :escape unless @input.wait_readable(0.02) && @input.getch == "["
      case @input.getch
      when "A" then :up
      when "B" then :down
      when "C" then :right
      when "D" then :left
      else :ignore
      end
    end

    def printable(char)
      byte = char&.bytes&.first
      byte && byte >= 0x20 && byte < 0x7F ? [:char, char] : :ignore
    end

    # --- helpers -------------------------------------------------------------

    def name_width(items)
      longest = items.map { |item| item.name.to_s.length }.max.to_i
      longest.clamp(NAME_MIN, NAME_MAX)
    end

    def truncate(value, width)
      return "" if width <= 0
      value.length > width ? "#{value[0, [width - 1, 0].max]}…" : value
    end

    def display_width(value)
      value.length
    end

    def columns
      winsize ? winsize[1] : 80
    end

    # The list fills whatever vertical space the popup gives us (minus chrome),
    # so the window grows with the popup and never leaves a gap below itself.
    def list_height
      rows = winsize ? winsize[0] : 24
      [rows - CHROME, MIN_LIST].max
    end

    def winsize
      (@output.respond_to?(:winsize) && @output.tty? && @output.winsize) || IO.console&.winsize
    end
  end
end

File: internal/palette/palette.go
package palette

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"os/signal"
	"sort"
	"strings"
	"time"

	"golang.org/x/sys/unix"
	"golang.org/x/term"
)

// escapeTimeout is how long readKey waits for the rest of an escape sequence
// (e.g. arrow keys send ESC [ A) before treating a bare ESC as "escape". This
// mirrors the Ruby palette's 20ms wait_readable; without it Peek would block on
// a lone ESC until the next keypress, so dismissing the menu took several Escs.
const escapeTimeout = 50 * time.Millisecond

type StateKind string

const (
	StateRunning     StateKind = "running"
	StateIdle        StateKind = "idle"
	StateFinished    StateKind = "finished"
	StateAvailable   StateKind = "available"
	StateDisabled    StateKind = "disabled"
	StateUnavailable StateKind = "unavailable"
)

type Item struct {
	Token       string
	Section     string
	Name        string
	Description string
	Command     string
	Directory   string
	State       string
	StateKind   StateKind
	Summary     string
}

type Selection struct {
	Action string
	Token  string
}

type Option struct {
	Token string
	Label string
}

const (
	MinList             = 3
	RightPad            = 2
	Reset               = "\033[0m"
	Bold                = "\033[1m"
	Dim                 = "\033[2m"
	Reverse             = "\033[7m"
	Header              = "\033[1;35m"
	// Accent is the ayu theme orange (#ff8f40) as a truecolor foreground, used
	// for the selected-row bar and the detail box border.
	Accent = "\033[38;2;255;143;64m"
	// Muted is the ayu surface_dim (#273747), a subtle grey for the divider
	// between the running sidebar and the searchable list.
	Muted               = "\033[38;2;39;55;71m"
	RunningSection      = "Running"
	SidebarWidth        = 20
	SummarySidebarWidth = 30
	NameMin             = 10
	NameMax             = 24
)

var StateColor = map[StateKind]string{
	StateRunning:     "\033[32m",
	StateFinished:    "\033[33m",
	StateAvailable:   "\033[32m",
	StateDisabled:    Dim,
	StateUnavailable: "\033[31m",
	StateIdle:        Dim,
}

type Palette struct {
	Input  *os.File
	Output io.Writer
	Prompt string
}

func New() *Palette {
	return &Palette{Input: os.Stdin, Output: os.Stdout, Prompt: "Muxpad"}
}

func (p *Palette) Select(items []Item, sectionOrder []string) (Selection, bool, error) {
	if p.Input == nil {
		p.Input = os.Stdin
	}
	if p.Output == nil {
		p.Output = os.Stdout
	}
	if !term.IsTerminal(int(p.Input.Fd())) {
		return Selection{}, false, fmt.Errorf("the Muxpad palette requires an interactive terminal")
	}
	running, launch := splitRunning(items)
	model := newModel(launch, running, sectionOrder)
	render := func() {
		columns, rows := terminalSize(p.Output)
		fmt.Fprint(p.Output, "\033[?25l\033[H")
		for i, line := range model.Render(columns-RightPad, rows) {
			if i > 0 {
				fmt.Fprint(p.Output, "\r\n")
			}
			fmt.Fprintf(p.Output, "%s\033[K", line)
		}
		fmt.Fprint(p.Output, "\033[J")
	}
	return p.interact(render, model.handle)
}

func (p *Palette) Choose(options []Option, title string) (string, bool, error) {
	if p.Input == nil {
		p.Input = os.Stdin
	}
	if p.Output == nil {
		p.Output = os.Stdout
	}
	if !term.IsTerminal(int(p.Input.Fd())) {
		return "", false, fmt.Errorf("the Muxpad palette requires an interactive terminal")
	}
	cursor := 0
	render := func() {
		fmt.Fprint(p.Output, "\033[?25l\033[H")
		fmt.Fprintf(p.Output, "%s  %s%s\033[K\r\n\r\n", Bold, title, Reset)
		for i, option := range options {
			line := "  " + option.Label
			if i == cursor {
				line = Reverse + line + Reset
			}
			fmt.Fprintf(p.Output, "%s\033[K\r\n", line)
		}
		fmt.Fprintf(p.Output, "\r\n  %senter select · esc cancel%s\033[J", Dim, Reset)
	}
	oldState, err := term.MakeRaw(int(p.Input.Fd()))
	if err != nil {
		return "", false, err
	}
	defer func() {
		_ = term.Restore(int(p.Input.Fd()), oldState)
		fmt.Fprint(p.Output, "\033[?25h\033[2J\033[H")
	}()
	handle := func(key string) (Selection, bool, bool) {
		switch key {
		case "up":
			if cursor > 0 {
				cursor--
			}
		case "down":
			if cursor < len(options)-1 {
				cursor++
			}
		case "enter":
			if len(options) == 0 {
				return Selection{}, false, true
			}
			return Selection{Token: options[cursor].Token}, true, true
		case "escape", "cancel":
			return Selection{}, false, true
		}
		return Selection{}, false, false
	}
	keys := p.readKeys()
	resize := watchResize()
	defer signal.Stop(resize)
	selection, ok := runLoop(render, handle, keys, resize)
	return selection.Token, ok, nil
}

func (p *Palette) interact(render func(), handle func(string) (Selection, bool, bool)) (Selection, bool, error) {
	oldState, err := term.MakeRaw(int(p.Input.Fd()))
	if err != nil {
		return Selection{}, false, err
	}
	defer func() {
		_ = term.Restore(int(p.Input.Fd()), oldState)
		fmt.Fprint(p.Output, "\033[?25h\033[2J\033[H")
	}()
	keys := p.readKeys()
	resize := watchResize()
	defer signal.Stop(resize)
	selection, ok := runLoop(render, handle, keys, resize)
	return selection, ok, nil
}

// readKeys pumps decoded keys onto a channel from a background reader. The
// palette process is short-lived, so a reader parked on Stdin after the loop
// returns is harmless.
func (p *Palette) readKeys() <-chan string {
	keys := make(chan string)
	reader := bufio.NewReader(p.Input)
	fd := int(p.Input.Fd())
	go func() {
		for {
			keys <- readKey(reader, fd)
		}
	}()
	return keys
}

// watchResize delivers SIGWINCH so the loop can repaint. Overlay panes (herdr
// overlays, tmux popups) often settle their size just after the child starts;
// without this the first paint is drawn at the wrong size and only corrected on
// the next keypress.
func watchResize() chan os.Signal {
	resize := make(chan os.Signal, 1)
	signal.Notify(resize, unix.SIGWINCH)
	return resize
}

// runLoop renders once, then repaints on every key and on every resize, ending
// when handle reports it is done. Splitting this out keeps the event handling
// testable without a real terminal.
func runLoop(render func(), handle func(string) (Selection, bool, bool), keys <-chan string, resize <-chan os.Signal) (Selection, bool) {
	render()
	for {
		select {
		case <-resize:
			render()
		case key, open := <-keys:
			if !open {
				return Selection{}, false
			}
			selection, ok, done := handle(key)
			if done {
				return selection, ok
			}
			render()
		}
	}
}

type model struct {
	running      []Item
	launch       []Item
	sectionOrder []string
	query        string
	focusRunning bool
	cursor       int
	runCursor    int
	rows         []row
	selectable   []int
	nameWidth    int
	offset       int
}

type row struct {
	kind  string
	label string
	item  Item
}

func newModel(launch, running []Item, sectionOrder []string) *model {
	m := &model{launch: launch, running: running, sectionOrder: sectionOrder, nameWidth: NameWidth(launch)}
	m.refilter()
	return m
}

func (m *model) handle(key string) (Selection, bool, bool) {
	switch key {
	case "up":
		m.move(-1)
	case "down":
		m.move(1)
	case "left":
		if len(m.running) > 0 {
			m.focusRunning = true
		}
	case "right":
		m.focusRunning = false
	case "backspace":
		if m.query != "" {
			runes := []rune(m.query)
			m.query = string(runes[:len(runes)-1])
			m.focusRunning = false
			m.refilter()
		}
	case "clear":
		m.query = ""
		m.focusRunning = false
		m.refilter()
	case "enter":
		if current, ok := m.current(); ok {
			return Selection{Action: "enter", Token: current.Token}, true, true
		}
	case "tab":
		if !m.focusRunning {
			if current, ok := m.current(); ok {
				return Selection{Action: "tab", Token: current.Token}, true, true
			}
		}
	case "restart":
		if current, ok := m.current(); ok && current.StateKind == StateFinished {
			return Selection{Action: "ctrl-r", Token: current.Token}, true, true
		}
	case "escape", "cancel":
		return Selection{}, false, true
	default:
		if strings.HasPrefix(key, "char:") {
			m.query += strings.TrimPrefix(key, "char:")
			m.focusRunning = false
			m.refilter()
		}
	}
	return Selection{}, false, false
}

func (m *model) refilter() {
	type rankedItem struct {
		item  Item
		score int
		index int
	}
	grouped := map[string][]rankedItem{}
	for index, item := range m.launch {
		if score := Score(item, m.query); score >= 0 {
			grouped[item.Section] = append(grouped[item.Section], rankedItem{item: item, score: score, index: index})
		}
	}
	m.rows = nil
	for _, section := range m.sectionOrder {
		items := grouped[section]
		if len(items) == 0 {
			continue
		}
		sort.SliceStable(items, func(i, j int) bool {
			if items[i].score == items[j].score {
				return items[i].index < items[j].index
			}
			return items[i].score > items[j].score
		})
		if len(m.rows) > 0 {
			m.rows = append(m.rows, row{kind: "blank"})
		}
		m.rows = append(m.rows, row{kind: "header", label: section})
		for _, item := range items {
			m.rows = append(m.rows, row{kind: "item", item: item.item})
		}
	}
	m.selectable = nil
	for i, row := range m.rows {
		if row.kind == "item" {
			m.selectable = append(m.selectable, i)
		}
	}
	m.cursor = 0
	m.offset = 0
}

func (m *model) current() (Item, bool) {
	if m.focusRunning {
		if len(m.running) == 0 {
			return Item{}, false
		}
		return m.running[m.runCursor], true
	}
	if len(m.selectable) == 0 {
		return Item{}, false
	}
	return m.rows[m.selectable[m.cursor]].item, true
}

func (m *model) move(delta int) {
	if m.focusRunning {
		m.runCursor = clamp(m.runCursor+delta, 0, len(m.running)-1)
		return
	}
	m.cursor = clamp(m.cursor+delta, 0, len(m.selectable)-1)
}

func (m *model) Render(totalWidth, totalRows int) []string {
	if totalWidth <= 0 {
		totalWidth = 80
	}
	top := m.topLines()
	footer := m.footerLines(totalWidth)
	listHeight := max(totalRows-len(top)-len(footer), MinList)
	body := m.BodyLines(totalWidth, listHeight)
	pad := max(totalRows-len(top)-len(body)-len(footer), 0)
	lines := append([]string{}, top...)
	lines = append(lines, body...)
	for range pad {
		lines = append(lines, "")
	}
	lines = append(lines, footer...)
	return lines
}

// topLines renders the filter input. The overlay pane already draws a titled
// border, so there is no in-palette heading (which used to duplicate it).
func (m *model) topLines() []string {
	input := "  " + Bold + "❯" + Reset + " "
	if m.query == "" {
		input += Dim + "Type to filter" + Reset
	} else {
		input += m.query
	}
	return []string{"", input, ""}
}

// footerLines renders the detail box followed by the key hint.
func (m *model) footerLines(totalWidth int) []string {
	lines := append([]string{""}, m.detailBox(totalWidth)...)
	return append(lines, "", "  "+Dim+m.hint()+Reset)
}

func (m *model) BodyLines(totalWidth, listHeight int) []string {
	if totalWidth <= 0 {
		totalWidth = 80
	}
	launch := m.launchLines(totalWidth, listHeight)
	if len(m.running) == 0 {
		return launch
	}
	sideWidth := SidebarWidth
	for _, item := range m.running {
		if item.Summary != "" {
			sideWidth = SummarySidebarWidth
			break
		}
	}
	launch = m.launchLines(totalWidth-sideWidth-1, listHeight)
	side := SidebarLines(m.running, sideWidth, m.focusRunning, m.runCursor)
	height := min(max(len(launch), len(side)), listHeight)
	out := make([]string, 0, height)
	for i := 0; i < height; i++ {
		left := strings.Repeat(" ", sideWidth)
		if i < len(side) {
			left = side[i]
		}
		right := ""
		if i < len(launch) {
			right = launch[i]
		}
		out = append(out, left+Muted+"│"+Reset+right)
	}
	return out
}

func (m *model) launchLines(width, height int) []string {
	m.clampOffset(height)
	last := min(m.offset+height, len(m.rows))
	out := make([]string, 0, last-m.offset)
	for _, row := range m.rows[m.offset:last] {
		switch row.kind {
		case "blank":
			out = append(out, "")
		case "header":
			out = append(out, "  "+Header+row.label+Reset)
		case "item":
			selected := false
			if current, ok := m.current(); ok && !m.focusRunning && current.Token == row.item.Token {
				selected = true
			}
			out = append(out, RenderItem(row.item, width, m.nameWidth, selected))
		}
	}
	return out
}

func (m *model) clampOffset(height int) {
	if len(m.selectable) == 0 {
		m.offset = 0
		return
	}
	line := m.selectable[m.cursor]
	if line < m.offset {
		m.offset = line
	}
	if line > m.offset+height-1 {
		m.offset = line - height + 1
	}
	m.offset = clamp(m.offset, 0, max(len(m.rows)-1, 0))
}

// detailBox renders a rounded, accent-bordered box previewing the current item.
// The box title is the verb (Run / Open / Focus); the body shows the command
// (for tasks/scripts) and the directory, each with an icon.
func (m *model) detailBox(totalWidth int) []string {
	item, ok := m.current()
	if !ok {
		return nil
	}
	boxWidth := max(totalWidth-2, 12)
	inner := boxWidth - 4
	var content []string
	if item.Command != "" && item.Command != item.Directory {
		content = append(content, boxContent(Accent+"▶"+Reset+" "+Truncate(item.Command, inner-2), inner))
	}
	if item.Directory != "" {
		content = append(content, boxContent("📁 "+Dim+Truncate(Abbreviate(item.Directory), inner-3)+Reset, inner))
	}
	if item.State != "" {
		color := StateColor[item.StateKind]
		if color == "" {
			color = Dim
		}
		content = append(content, boxContent(color+stateGlyph(item.StateKind)+" "+Truncate(item.State, inner-2)+Reset, inner))
	}
	if len(content) == 0 {
		content = append(content, boxContent("", inner))
	}
	title := " " + m.verb() + " "
	dashes := max(boxWidth-3-visibleWidth(title), 0)
	top := "  " + Accent + "╭─" + Bold + title + Reset + Accent + strings.Repeat("─", dashes) + "╮" + Reset
	bottom := "  " + Accent + "╰" + strings.Repeat("─", max(boxWidth-2, 0)) + "╯" + Reset
	lines := []string{top}
	for _, c := range content {
		lines = append(lines, "  "+Accent+"│"+Reset+" "+c+" "+Accent+"│"+Reset)
	}
	return append(lines, bottom)
}

func (m *model) verb() string {
	item, ok := m.current()
	if !ok {
		return "Run"
	}
	if item.StateKind == StateRunning || item.StateKind == StateFinished {
		return "Focus"
	}
	// Project entries carry a directory but no command; they open a workspace.
	if item.Command == "" && item.Directory != "" {
		return "Open"
	}
	return "Run"
}

func (m *model) hint() string {
	switcher := ""
	if len(m.running) > 0 {
		switcher = "←/→ switch · "
	}
	current, ok := m.current()
	restartable := ok && current.StateKind == StateFinished
	if m.focusRunning {
		actions := []string{"enter focus"}
		if restartable {
			actions = append(actions, "ctrl-r restart")
		}
		return switcher + strings.Join(actions, " · ") + " · esc close"
	}
	actions := []string{"enter " + strings.ToLower(m.verb())}
	// Placement actions apply to launchable items (which carry a command), not
	// to project entries.
	if ok && current.Command != "" {
		actions = append(actions, "tab actions")
	}
	if restartable {
		actions = append(actions, "ctrl-r restart")
	}
	return switcher + strings.Join(actions, " · ") + " · esc close"
}

func splitRunning(items []Item) ([]Item, []Item) {
	var running, launch []Item
	for _, item := range items {
		if item.Section == RunningSection {
			running = append(running, item)
		} else {
			launch = append(launch, item)
		}
	}
	return running, launch
}

func readKey(reader *bufio.Reader, fd int) string {
	r, _, err := reader.ReadRune()
	if err != nil {
		return "escape"
	}
	switch r {
	case '\r', '\n':
		return "enter"
	case '\t':
		return "tab"
	case 0x7f, '\b':
		return "backspace"
	case 0x03:
		return "cancel"
	case 0x12:
		return "restart"
	case 0x0e:
		return "down"
	case 0x10:
		return "up"
	case 0x15:
		return "clear"
	case 0x1b:
		// Wait briefly for the rest of an escape sequence. A bare ESC has no
		// follow-up bytes, so the poll times out and we treat it as "escape"
		// without blocking on the next keypress.
		if !waitReadable(reader, fd, escapeTimeout) {
			return "escape"
		}
		if b, err := reader.ReadByte(); err != nil || b != '[' {
			return "escape"
		}
		dir, err := reader.ReadByte()
		if err != nil {
			return "escape"
		}
		switch dir {
		case 'A':
			return "up"
		case 'B':
			return "down"
		case 'C':
			return "right"
		case 'D':
			return "left"
		}
		return "ignore"
	default:
		if r >= 0x20 && r < 0x7f {
			return "char:" + string(r)
		}
		return "ignore"
	}
}

// waitReadable reports whether more input is available within timeout, either
// already buffered by the reader or arriving on fd. It retries on EINTR so a
// signal does not spuriously report "no input".
func waitReadable(reader *bufio.Reader, fd int, timeout time.Duration) bool {
	if reader.Buffered() > 0 {
		return true
	}
	fds := []unix.PollFd{{Fd: int32(fd), Events: unix.POLLIN}}
	ms := int(timeout.Milliseconds())
	for {
		n, err := unix.Poll(fds, ms)
		if err == unix.EINTR {
			continue
		}
		return err == nil && n > 0
	}
}

func terminalSize(output io.Writer) (int, int) {
	if file, ok := output.(*os.File); ok && term.IsTerminal(int(file.Fd())) {
		w, h, err := term.GetSize(int(file.Fd()))
		if err == nil {
			return w, h
		}
	}
	return 80, 24
}

func clamp(value, low, high int) int {
	if high < low {
		return low
	}
	if value < low {
		return low
	}
	if value > high {
		return high
	}
	return value
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
